File: rust/src/day13.rs
/// Day 13: Mine Cart Madness
/// 
/// Parse and trace mine-cart tracks and figure out when they'll collide

// Note that +x is right and +y is down

#[derive(Clone)]
pub enum Turn {
    Left,
    Straight,
    Right,
}

#[derive(Clone)]
pub struct Cart {
    x: usize,
    y: usize,
    next_turn: Turn,
    direction: (isize, isize),
}

pub fn parse_map(map: &str) -> (Vec<Vec<char>>, Vec<Cart>) {
    let mut positions: Vec<Vec<char>> = vec![];
    let mut carts: Vec<Cart> = vec![];

    for (y, line) in map.lines().enumerate() {
        positions.push(vec![]);
        for (x, c) in line.chars().enumerate() {
            if c == '>' {
                let cart = Cart { x, y, next_turn: Turn::Left, direction: (1, 0)};
                carts.push(cart);
                positions[y].push('-');
            } else if c == '<' {
                let cart = Cart { x, y, next_turn: Turn::Left, direction: (-1, 0) };
                carts.push(cart);
                positions[y].push('-');
            } else if c == '^' {
                let cart = Cart { x, y, next_turn: Turn::Left, direction: (0, -1) };
                carts.push(cart);
                positions[y].push('|');
            } else if c == 'v' {
                let cart = Cart { x, y, next_turn: Turn::Left, direction: (0, 1) };
                carts.push(cart);
                positions[y].push('|');
            } else {
                positions[y].push(c);
            }
        }
    }
    (positions, carts)
}

pub fn find_crash(map: &str) -> (usize, usize) {
    let (tracks, mut carts) = parse_map(map);
    loop {
        carts.sort_by_key(|cart| (cart.y, cart.x));

        for i in 0..carts.len() {
            {
                let mut cart = &mut carts[i];
                let (x, y) = cart.direction;
                cart.x = (cart.x as isize + x) as usize;
                cart.y = (cart.y as isize + y) as usize;
                let location = tracks[cart.y][cart.x];
                if location == '+' {
                    match cart.next_turn {
                        Turn::Left => {
                            cart.direction = turn_left(cart.direction);
                            cart.next_turn = Turn::Straight;
                        },
                        Turn::Straight => {
                            cart.next_turn = Turn::Right;
                        },
                        Turn::Right => {
                            cart.direction = turn_right(cart.direction);
                            cart.next_turn = Turn::Left;
                        }
                    }
                } else if location == '/' {
                    match cart.direction {
                        (0, -1) => cart.direction = (1, 0),
                        (-1, 0) => cart.direction = (0, 1),
                        (1, 0) => cart.direction = (0, -1),
                        (0, 1) => cart.direction = (-1, 0),
                        _ => panic!("Impossible direction"),
                    }
                } else if location == '\\' {
                    match cart.direction {
                        (1, 0) => cart.direction = (0, 1),
                        (0, 1) => cart.direction = (1, 0),
                        (-1, 0) => cart.direction = (0, -1),
                        (0, -1) => cart.direction = (-1, 0),
                        _ => panic!("Impossible direction"),
                    }
                }
            }    
            let cart = &carts[i];
            if carts.iter().filter(|c| c.x == cart.x && c.y == cart.y).count() > 1 {
                return (cart.x, cart.y);
            }
        }
    }
}

fn turn_left(direction: (isize, isize)) -> (isize, isize) {
    if direction == (1, 0) { (0, -1) }
    else if direction == (0, -1) { (-1, 0) }
    else if direction == (-1, 0) { (0, 1) }
    else { (1, 0) }
}

fn turn_right(direction: (isize, isize)) -> (isize, isize) {
    if direction == (1, 0) { (0, 1) }
    else if direction == (0, 1) { (-1, 0) }
    else if direction == (-1, 0) { (0, -1) }
    else { (1, 0) }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_part_one() {
        let map = "/->-\\        
|   |  /----\\
| /-+--+-\\  |
| | |  | v  |
\\-+-/  \\-+--/
  \\------/   ";
    
        assert_eq!((7, 3), find_crash(map));
    }
}

File: python/src/day13.py
"""Day 13: Mine Cart Madness

Figure out when carts on tracks will crash
"""

from typing import List


class Vector:
    """A 2D vector.  Right is +X, Down is +Y"""

    def __init__(self, x, y):
        self.x = x
        self.y = y

    def __add__(self, other):
        return Vector(self.x + other.x, self.y + other.y)

    def __eq__(self, other):
        return self.x == other.x and self.y == other.y

    def __repr__(self):
        return f"Vector({self.x}, {self.y})"

    def __str__(self):
        return f"<{self.x}, {self.y}>"

    def __hash__(self):
        return hash((self.x, self.y))


NORTH = Vector(0, -1)
SOUTH = Vector(0, 1)
EAST = Vector(1, 0)
WEST = Vector(-1, 0)
DIRECTIONS = [NORTH, EAST, SOUTH, WEST]


class Cart:
    """A cart that drives around the mine"""

    def __init__(self, position: Vector, direction: Vector):
        self.position = position
        self.direction = direction
        self.next_turn = "left"

    def move(self):
        """Updates position based on direction"""
        self.position += self.direction

    def turn_right(self):
        self.direction = DIRECTIONS[
            (DIRECTIONS.index(self.direction) + 1) % len(DIRECTIONS)
        ]

    def turn_left(self):
        self.direction = DIRECTIONS[
            (DIRECTIONS.index(self.direction) - 1) % len(DIRECTIONS)
        ]

    def handle_next_turn(self):
        """Update own direction based on the next turn value specified in the prompt"""
        if self.next_turn == "left":
            self.turn_left()
            self.next_turn = "straight"
        elif self.next_turn == "straight":
            self.next_turn = "right"
        elif self.next_turn == "right":
            self.turn_right()
            self.next_turn = "left"


class Mine:
    """A mine that has a bunch of different carts on tracks"""

    def __init__(self, map: str):
        self.map = []
        self.carts: List[Cart] = []
        for y, line in enumerate(map.splitlines()):
            self.map.append([])
            for x, c in enumerate(line):
                if c == '>':
                    self.carts.append(Cart(Vector(x, y), EAST))
                    self.map[-1].append('-')
                elif c == '<':
                    self.carts.append(Cart(Vector(x, y), WEST))
                    self.map[-1].append('-')
                elif c == '^':
                    self.carts.append(Cart(Vector(x, y), NORTH))
                    self.map[-1].append('|')
                elif c == 'v':
                    self.carts.append(Cart(Vector(x, y), SOUTH))
                    self.map[-1].append('|')
                else:
                    self.map[-1].append(c)

    def find_first_collision(self) -> Vector:
        """Simulates the mine until two carts collide.  Returns that x, y pair"""
        while True:
            self.carts.sort(key=lambda cart: (
                cart.position.y, cart.position.x))

            for cart in self.carts:
                cart.move()
                self.turn_cart(cart)

                if self.is_collision(cart):
                    return cart.position

    def last_cart_location(self) -> Vector:
        """Simulates the mine, removing carts when they collide.
        Returns the x, y pair of the last cart standing
        """
        while True:
            self.sort_carts()

            crashed = set()
            for cart in self.carts:
                if cart.position in crashed:
                    continue
                cart.move()
                self.turn_cart(cart)

                if self.is_collision(cart):
                    crashed.add(cart.position)
                    continue

            self.carts = [cart
                          for cart in self.carts
                          if cart.position not in crashed]
            if len(self.carts) == 1:
                return self.carts[0].position

    def sort_carts(self):
        """Carts are always run in reading order, top to bottom, left to right"""
        self.carts.sort(key=lambda cart: (cart.position.y, cart.position.x))

    def is_collision(self, cart: Cart) -> bool:
        """Checks whether or not a cart is running into any other carts"""
        return sum(other.position == cart.position for other in self.carts) > 1

    def turn_cart(self, cart: Cart):
        """Depending on what kind of track the cart is on, turn it appropriately"""
        track_type = self.map[cart.position.y][cart.position.x]
        if track_type == '+':
            cart.handle_next_turn()
        elif track_type == '/':
            if cart.direction == NORTH or cart.direction == SOUTH:
                cart.turn_right()
            else:
                cart.turn_left()
        elif track_type == '\\':
            if cart.direction == NORTH or cart.direction == SOUTH:
                cart.turn_left()
            else:
                cart.turn_right()


if __name__ == "__main__":
    with open("python/data/day13.txt", "r") as f:
        mine_map = f.read()

    # Part 1
    mine = Mine(mine_map)
    print(mine.find_first_collision())

    # Part 2
    mine = Mine(mine_map)
    print(mine.last_cart_location())
